Documentation/proposals/202304-agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,151 @@
+
+Looking at the traditional deployment model, we have a single Prometheus (or an HA setup) per cluster or namespace, responsible for scraping all containers under their scope. Prometheus operator relies on `ServiceMonitor`, `PodMonitor`, and `Probe` CRs to configure Prometheus, which will eventually use Kubernetes service-discovery to find endpoints that need to be scraped.
+
+### Pitfalls of the current solution
+
+Depending on the Cluster's scale and how often Prometheus hits Kubernetes API, Prometheus service discovery can increase the load on the API significantly and affect the overall functionality of said cluster.
+
+Another problem is that one or more containers can be updated to a problematic version that causes a [Cardinality Spike](https://grafana.com/blog/2022/02/15/what-are-cardinality-spikes-and-why-do-they-matter/). Depending on the proportion of the spike, it is possible that a container could single-handedly crash the monitoring system of the whole cluster.
+
+![Traditional Deployment Pattern](assets/agent-deployment-pattern-sidecar/traditional-deployment-pattern.png "Traditional Deployment Pattern").
+
+## Goals
+
+* Allow better horizontal scalability of Prometheus.
+* Provide new deployment strategies for SRE teams.
+
+### Audience
+
+* SRE/Platform teams that are responsible for managing Prometheus instances.
+* Users who are looking for ways to horizontally scale Prometheus.
+
+## Non-Goals
+
+* Implement other deployment patterns, e.g. daemonset.
+
+## How
+
+This document proposes a new deployment model where Prometheus-Operator injects Prometheus agents as a sidecar container (and Prometheus config reloader) to pods that need to be scrapped. With a sidecar, we tackle both problems mentioned above:
+
+> * Load on Kubernetes API won't exist since it's not needed anymore. Prometheus will scrape containers from the same pod through their shared network interface and scrape configuration can be declared via pod annotations.
+> * A sudden cardinality spike will not affect the whole monitoring system. In a worst-case scenario, it will fail a single pod.
+
+The existing PrometheusAgent CRD would be extended with a new field called `mode`, which can be one of two values(for now): [`statefulset`, `sidecar`], with `statefulset` as default. If `mode` is set to `sidecar`, Prometheus-Operator won't deploy any Prometheus agents initially. Instead, it will use `ServiceMonitor`, `PodMonitor`, and `Probes` to find pods that need to be scrapped through then inject Prometheus Agent as a sidecar to the pods found through the label selectors found in the mentioned resources.
```

Comment:
More details on how the operator will know which resources it should reconcile and how they will materialize in the resulting Prometheus config?

Reply:
That's a good point, we probably want a different API for the Status, I still need to think about how it should look